File: src/ca/uqam/mgl7460/tp1/tests/TestCreationInstanceProcessus.java
package ca.uqam.mgl7460.tp1.tests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeAll;

import ca.uqam.mgl7460.tp1.types.modeles.DemandePret;
import ca.uqam.mgl7460.tp1.types.modeles.Resultat;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionProcessus;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionTache;
import ca.uqam.mgl7460.tp1.types.traitements.instances.ExceptionDefinitionProcessus;
import ca.uqam.mgl7460.tp1.types.traitements.instances.InstanceProcessus;
import ca.uqam.mgl7460.tp1.types.utils.Fabrique;

public class TestCreationInstanceProcessus {

    private static DefinitionProcessus processusLineaire = null;

    private static DefinitionProcessus processusComplexe = null;

    private static DefinitionProcessus processusErrone = null;

    private static Fabrique fabrique = null;

    @BeforeAll
    public static void initialiser() {

        fabrique = Fabrique.getSingletonFabrique();

        processusLineaire = CreationDefinitionsProcessus.getProcessusLineaire();

        processusComplexe = CreationDefinitionsProcessus.getProcessusComplexe();

        processusErrone = CreationDefinitionsProcessus.getProcessusErrone();
    }

    @Test
    public void testCreationInstanceProcessus() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiOui();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);

        assertEquals(processusLineaire,instanceProcessus.getDefinitionProcessus(),"Le nouveau processus n'a pas la bonne definition");
    }

    @Test
    public void testExecutionProcessusSimpleSurDemandeOuiOuiOui() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiOui();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);
        try {
            instanceProcessus.demarrer();
            assertEquals(Resultat.ACCEPTEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être acceptée");
        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }
    }

    @Test void testExecutionProcessusSimpleSurDemandeOuiOuiNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);
        try {
            instanceProcessus.demarrer();

            // la demande est supposée être refusée
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");
            // et avoir échoué la troisième tâche
            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibilitePret()));

        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }
    }

    private boolean aEchoueLesTaches(DemandePret demandePret, DefinitionTache...definitionTaches){

        ArrayList<String> messagesErreur = new ArrayList<>();
        demandePret.getResultatTraitement().getMessages().forEachRemaining(message -> messagesErreur.add(message));

        if (messagesErreur.size() != definitionTaches.length) return false;

        boolean itMatchesAll = true;
        for (DefinitionTache tache: definitionTaches) {
            String nomTache = tache.getNom();
            boolean itMatchesTask = false;
            for (String message: messagesErreur) {
                itMatchesTask = itMatchesTask || message.contains(nomTache);
            }
            itMatchesAll = itMatchesAll && itMatchesTask;
        }
        return itMatchesAll;
    }

    @Test 
    void testExecutionProcessusSimpleSurDemandeOuiNonNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiNonNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);
        try {
            instanceProcessus.demarrer();
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");

            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibilitePropriete() ,CreationDefinitionsTaches.getTacheEligibilitePret()));
        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }

    }
    @Test 
    void testExecutionProcessusSimpleSurDemandeNonNonNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretNonNonNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);
        try {
            instanceProcessus.demarrer();
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");

            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibiliteEmprunteur() ,CreationDefinitionsTaches.getTacheEligibilitePropriete() ,CreationDefinitionsTaches.getTacheEligibilitePret()));
        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }

    }

    
    @Test 
    void testExecutionProcessusSimpleSurDemandeOuiNonOui() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiNonOui();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusLineaire, demandePret);
        try {
            instanceProcessus.demarrer();
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");
            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibilitePropriete()));
        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }

    }

    @Test
    void testExecutionProcessusMalForme(){
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiOui();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusErrone, demandePret);

        assertThrows(ExceptionDefinitionProcessus.class, () -> {instanceProcessus.demarrer();});
    }

    @Test
    void testExecutionProcessusComplexeSurDemandeOuiOuiOui() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiOui();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusComplexe, demandePret);
        try {
            instanceProcessus.demarrer();
            assertEquals(Resultat.ACCEPTEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être acceptée");

        } catch (ExceptionDefinitionProcessus edp) {
            edp.printStackTrace();
        }

    }

     @Test void testExecutionProcessusComplexeSurDemandeOuiOuiNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiOuiNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusComplexe, demandePret);
        try {
            instanceProcessus.demarrer();

            // la demande est supposée être refusée
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");
            // et avoir échoué la troisième tâche
            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibilitePret()));

        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }
    }

         @Test void testExecutionProcessusComplexeSurDemandeOuiNonNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretOuiNonNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusComplexe, demandePret);
        try {
            instanceProcessus.demarrer();

            // la demande est supposée être refusée
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");
            // et avoir échoué la deuxième tâche, auquel cas le processus se termine, de toutes les
            // façons
            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibilitePropriete()));

        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }
    }

        @Test 
        void testExecutionProcessusComplexeSurDemandeNonNonNon() {
        DemandePret demandePret = CreationDemandesPret.getDemandePretNonNonNon();
        InstanceProcessus instanceProcessus = fabrique.creerInstanceProcessus(processusComplexe, demandePret);
        try {
            instanceProcessus.demarrer();

            // la demande est supposée être refusée
            assertEquals(Resultat.REFUSEE,demandePret.getResultatTraitement().getResultat(), "Demande " + demandePret + " était supposée être refusée au niveau du pret");
            // et avoir échoué la première tâche, auquel cas le processus se termine, de toutes les
            // façons
            assertTrue(aEchoueLesTaches(demandePret,CreationDefinitionsTaches.getTacheEligibiliteEmprunteur()));

        } catch (ExceptionDefinitionProcessus e) {
            e.printStackTrace();
        }
    }


}

File: src_python/tests/test_creation_instance_processus.py
import unittest
from utils import fabrique
from tests.creation_definitions_processus import processus_lineaire, processus_errone, processus_complexe
from tests.creation_demandes_pret import demande_pret_oui_oui_oui, demande_pret_oui_oui_non, demande_pret_oui_non_non, demande_pret_non_non_non, demande_pret_oui_non_oui
from tests.creation_definitions_taches import tache_eligibilite_pret, tache_eligibilite_propriete, tache_eligibilite_emprunteur
from modeles.resultat import Resultat
from modeles.demande_pret import DemandePret
from traitements.instances.exception_definition_processus import ExceptionDefinitionProcessus
import traceback

class TestCreationInstanceProcessus(unittest.TestCase):

    def setUp(self):
        self.processus_lineaire = processus_lineaire()
        self.processus_errone = processus_errone()
        self.processus_complexe = processus_complexe()
    
    def test_creation_instance_processus(self):
        demande_pret = demande_pret_oui_oui_oui()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        self.assertEqual(self.processus_lineaire, instance_processus.definition_processus, "Le nouveau processus n'a pas la bonne definition")

    def test_execution_processus_simple_sur_demande_oui_oui_oui(self):
        demande_pret = demande_pret_oui_oui_oui()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.ACCEPTEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être acceptée")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_simple_sur_demande_oui_oui_non(self):
        demande_pret = demande_pret_oui_oui_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_pret() ), "La tache EligibilitePret était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def aEchoueLesTaches(self, demande_pret: DemandePret, *definitions_tache):
        errorMessages = demande_pret.resultat_traitement.messages

        if (len(errorMessages) != len(definitions_tache)): 
            return False
        
        isMatche = True
        for dt in definitions_tache:
            if not any(dt.nom in message for message in errorMessages):
                isMatche = False
                break
        return isMatche
    
    def test_execution_processus_simple_sur_demande_oui_non_non(self):
        demande_pret = demande_pret_oui_non_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_propriete(), tache_eligibilite_pret() ), "La tache EligibilitePropriete était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_simple_sur_demande_non_non_non(self):
        demande_pret = demande_pret_non_non_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_propriete(), tache_eligibilite_pret(), tache_eligibilite_emprunteur() ), "Les 3 taches étaient supposées échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_simple_sur_demande_oui_non_oui(self):
        demande_pret = demande_pret_oui_non_oui()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_lineaire, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_propriete()), "La tache EligibilitePropriete était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_mal_forme(self):
        demande_pret = demande_pret_oui_oui_oui()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_errone, demandePret=demande_pret)
        with self.assertRaises(ExceptionDefinitionProcessus):
            instance_processus.demarrer()
    
    def test_execution_processus_complexe_sur_demande_oui_oui_oui(self):
        demande_pret = demande_pret_oui_oui_oui()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_complexe, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.ACCEPTEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être acceptée")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_complexe_sur_demande_oui_oui_non(self):
        demande_pret = demande_pret_oui_oui_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_complexe, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_pret()), "La tache EligibilitePret était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()

    def test_execution_processus_complexe_sur_demande_oui_non_non(self):
        demande_pret = demande_pret_oui_non_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_complexe, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_propriete()), "La tache EligibilitePropriete était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()
    
    def test_execution_processus_complexe_sur_demande_non_non_non(self):
        demande_pret = demande_pret_non_non_non()
        instance_processus = fabrique.creer_instance_processus(definition_processus=self.processus_complexe, demandePret=demande_pret)
        try:
            instance_processus.demarrer()
            self.assertEqual(Resultat.REFUSEE, demande_pret.resultat_traitement.resultat, "Demande " + str(demande_pret) + " était supposée être refusée")
            self.assertTrue(self.aEchoueLesTaches(demande_pret, tache_eligibilite_emprunteur() ), "La tache EligibiliteEmprunteur était supposée échouer (aEchoueLesTaches failed)")
        except ExceptionDefinitionProcessus as e:
            traceback.print_exc()

if __name__ == '__main__':
    unittest.main()